File: web/src/protocol.ts
export type Judgment = "critical" | "perfect" | "great" | "good" | "miss";

export interface Score {
  combo: number;
  critical: number;
  perfect: number;
  great: number;
  good: number;
  miss: number;
  mine?: number;
  accuracy: number;
  achievement: number;
  dx_score: number;
  dx_max: number;
}

export interface SlideInfo {
  shape?: string;
  end_sensor?: string;
  path?: string[];
  end_t?: number;
  wait_t?: number;
  mid?: number | null;
}

export interface ChartNote {
  t: number;
  button: number | null;
  type: string;
  end?: number | null;
  sensor?: string;
  slide?: SlideInfo | null;
  is_break?: boolean;
  is_ex?: boolean;
  is_mine?: boolean;
  is_hanabi?: boolean;
  is_star?: boolean;
  is_each?: boolean;
  head_style?: string;
}

export interface ActiveNote {
  t: number;
  button: number | null;
  progress: number;
  sensor?: string;
  type?: string;
  end?: number | null;
  hold_phase?: "approach" | "sustain";
  path?: string[];
  slide?: SlideInfo | null;
  is_break?: boolean;
  is_ex?: boolean;
  is_mine?: boolean;
  is_hanabi?: boolean;
  is_star?: boolean;
  is_each?: boolean;
  head_style?: string;
}

export interface Drive {
  loomL: number;
  loomR: number;
  chaseL: number;
  chaseR: number;
  threatL: number;
  threatR: number;
}

export interface Pose {
  aim: number;
  strike: number;
  strike_l?: number;
  strike_r?: number;
  omega_l?: number;
  omega_r?: number;
  reach_l?: number;
  reach_r?: number;
}

export interface HelloMessage {
  type: "hello";
  version: number;
}

export interface ChartMessage {
  type: "chart";
  title: string;
  artist: string;
  notes: ChartNote[];
  active?: ActiveNote[];
  active_sensors?: string[];
  warnings?: number;
}

export interface BrainLayoutMessage {
  type: "brain_layout";
  neurons: number;
  x: number[];
  y: number[];
  groups: Record<string, number[]>;
  labels: Record<string, string>;
}

export interface StateMessage {
  type: "state";
  t: number;
  aim_button: number | null;
  tap: boolean;
  tap_button: number | null;
  aim_sensor?: string | null;
  tap_sensor?: string | null;
  hand_l_sensor?: string | null;
  hand_r_sensor?: string | null;
  /** Unit disk, x right, y up, origin C. Tip target; not a pad id. */
  hand_l?: [number, number] | null;
  hand_r?: [number, number] | null;
  score: Score;
  active: ActiveNote[];
  active_sensors?: string[];
  drive: Drive;
  pose: Pose;
  spikes?: number[];
  spike_total?: number;
}

export interface HitMessage {
  type: "hit";
  button: number | null;
  judgment: Judgment;
  sensor?: string | null;
  t: number;
  timing?: "fast" | "late" | null;
}

export interface EndMessage {
  type: "end";
  score: Score;
}

export interface ReadyMessage {
  type: "ready";
  message?: string;
  warnings?: number;
}

export interface LevelsMessage {
  type: "levels";
  levels: Array<{ difficulty: number; level: string }>;
}

export interface ErrorMessage {
  type: "error";
  message: string;
  warnings?: number;
}

export interface Resources {
  cpu_pct: number;
  sys_cpu_pct: number | null;
  rss_mb: number;
  ram_used_mb: number | null;
  ram_total_mb: number | null;
  gpu_pct: number | null;
  vram_used_mb: number | null;
  vram_total_mb: number | null;
  gpu_name: string | null;
}

export interface ResourcesMessage extends Resources {
  type: "resources";
}

export type ServerMessage =
  | HelloMessage
  | ChartMessage
  | BrainLayoutMessage
  | StateMessage
  | HitMessage
  | EndMessage
  | ReadyMessage
  | LevelsMessage
  | ErrorMessage
  | ResourcesMessage;

export type ClientMessage =
  | { type: "inspect_chart"; maidata: string }
  | { type: "load_chart"; maidata: string; difficulty?: number | null }
  | { type: "stop" };

export function isServerMessage(value: unknown): value is ServerMessage {
  if (!value || typeof value !== "object") return false;
  const type = (value as { type?: unknown }).type;
  return (
    type === "hello" ||
    type === "chart" ||
    type === "brain_layout" ||
    type === "state" ||
    type === "hit" ||
    type === "end" ||
    type === "ready" ||
    type === "levels" ||
    type === "error" ||
    type === "resources"
  );
}

File: fly_rg/protocol.py
"""WebSocket JSON message builders (shared with the web client)."""

from __future__ import annotations

import json
from typing import Any

from fly_rg.schema import Chart


def hello_message(version: int = 1) -> dict[str, Any]:
    return {"type": "hello", "version": version}


def chart_message(
    chart: Chart,
    *,
    active: list | None = None,
    active_sensors: list | None = None,
) -> dict[str, Any]:
    d = chart.to_dict()
    msg: dict[str, Any] = {
        "type": "chart",
        "title": d["title"],
        "artist": d["artist"],
        "notes": d["notes"],
    }
    if active is not None:
        msg["active"] = active
    if active_sensors is not None:
        msg["active_sensors"] = active_sensors
    return msg


def state_message(
    *,
    t: float,
    aim_button: int | None,
    tap: bool,
    tap_button: int | None,
    score: dict[str, Any],
    active: list[dict[str, Any]],
    drive: dict[str, float],
    pose: dict[str, float],
    spikes: list[int] | None = None,
    spike_total: int = 0,
    aim_sensor: str | None = None,
    tap_sensor: str | None = None,
    active_sensors: list[str] | None = None,
    hand_l_sensor: str | None = None,
    hand_r_sensor: str | None = None,
    hand_l: list[float] | tuple[float, float] | None = None,
    hand_r: list[float] | tuple[float, float] | None = None,
) -> dict[str, Any]:
    hl = [0.0, 0.0] if hand_l is None else [float(hand_l[0]), float(hand_l[1])]
    hr = [0.0, 0.0] if hand_r is None else [float(hand_r[0]), float(hand_r[1])]
    return {
        "type": "state",
        "t": float(t),
        "aim_button": aim_button,
        "aim_sensor": aim_sensor,
        "tap": bool(tap),
        "tap_button": tap_button,
        "tap_sensor": tap_sensor,
        "hand_l": hl,
        "hand_r": hr,
        "hand_l_sensor": hand_l_sensor,
        "hand_r_sensor": hand_r_sensor,
        "score": score,
        "active": active,
        "active_sensors": list(active_sensors or []),
        "drive": {
            "loomL": float(drive.get("loomL", 0.0)),
            "loomR": float(drive.get("loomR", 0.0)),
            "chaseL": float(drive.get("chaseL", 0.0)),
            "chaseR": float(drive.get("chaseR", 0.0)),
            "threatL": float(drive.get("threatL", 0.0)),
            "threatR": float(drive.get("threatR", 0.0)),
        },
        "pose": {
            "aim": float(pose.get("aim", 0.0)),
            "strike": float(pose.get("strike", 0.0)),
            "strike_l": float(pose.get("strike_l", pose.get("strike", 0.0))),
            "strike_r": float(pose.get("strike_r", pose.get("strike", 0.0))),
            "omega_l": float(pose.get("omega_l", 0.0)),
            "omega_r": float(pose.get("omega_r", 0.0)),
            "reach_l": float(pose.get("reach_l", 0.0)),
            "reach_r": float(pose.get("reach_r", 0.0)),
        },
        "spikes": list(spikes or []),
        "spike_total": int(spike_total),
    }


def brain_layout_message(layout: dict[str, Any]) -> dict[str, Any]:
    """Pass-through helper; layout already includes type=brain_layout."""
    return layout


def hit_message(
    *,
    t: float,
    button: int | None,
    judgment: str,
    sensor: str | None = None,
    timing: str | None = None,
) -> dict[str, Any]:
    return {
        "type": "hit",
        "t": float(t),
        "button": button,
        "sensor": sensor,
        "judgment": judgment,
        "timing": timing,
    }


def end_message(score: dict[str, Any]) -> dict[str, Any]:
    return {"type": "end", "score": score}


def resources_message(stats: dict[str, Any]) -> dict[str, Any]:
    def _num(key: str) -> float | None:
        value = stats.get(key)
        if value is None:
            return None
        return float(value)

    name = stats.get("gpu_name")
    gpu_name = None
    if isinstance(name, str):
        cleaned = "".join(ch for ch in name if ch.isprintable()).strip()
        gpu_name = cleaned[:64] if cleaned else None
    return {
        "type": "resources",
        "cpu_pct": float(stats.get("cpu_pct") or 0.0),
        "sys_cpu_pct": _num("sys_cpu_pct"),
        "rss_mb": float(stats.get("rss_mb") or 0.0),
        "ram_used_mb": _num("ram_used_mb"),
        "ram_total_mb": _num("ram_total_mb"),
        "gpu_pct": _num("gpu_pct"),
        "vram_used_mb": _num("vram_used_mb"),
        "vram_total_mb": _num("vram_total_mb"),
        "gpu_name": gpu_name,
    }


def dumps(message: dict[str, Any]) -> str:
    return json.dumps(message, separators=(",", ":"))
